### ISSUE TITLE
Impossible to see contents of AppArmor Profile "docker-default"

### ISSUE THREAD

**Issue by spawnflagger** [moby/moby#33060]
**Description**
I would like to be able to see the plaintext contents of the AppArmor profile "docker-default", for debugging/audit purposes.

This bug/feature-request is related to #24786 (fix implemented in #26518), and to a lesser extent #26823, #25935 (claimed to be fixed in #27083, however in my 17.04 install 'docker-inspect' still doesn't show the new behavior)

Prior to Docker 1.13, it stored the AppArmor Profile in /etc/apparmor.d/docker-default (which was overwritten when Docker started, so users couldn't modify it.  Docker devs added the --security-opt to let users specify a profile.  After v1.13, Docker now generates docker-default in tmpfs, uses apparmor_parser to load it into kernel, _then deletes the file_.  All of the AppArmor utils (aa-* on Ubuntu) expect a file parameter, and /sys/kernel/security/apparmor/policy/profiles/* only has cached binaries.

In discussion of #24786, @cyphar mentions this being opaque, and I agree.

**Steps to reproduce the issue:**
1.  `docker run` a container (without --privileged)
2. `aa-status` to confirm "docker-default" profile is being enforced
3. try to find what "docker-default" profile actually is, and the Policy/Policies being enforced

**Describe the results you received:**
failed to find information in step 3.

**Describe the results you expected:**
expected to find exactly what is being enforced.  (without reading the source code and running a Go template parser)

**Additional information you deem important (e.g. issue happens only occasionally):**
This came up as I was trying to debug a Docker container (Rails application based on phusion/passenger-ruby) because kern.log/syslog had excessive AppArmor DENIED ptrace for a 'ps', which according to Docker documentation should be allowed in the default profile. 

I apologize in advance, if there is already some way to get this info. (I'm new to Docker and AppArmor, but old to Linux/Solaris/AIX/BSD/etc)

Possible solution/suggestions:
- restore `/etc/apparmor.d/default-docker`, and only change it during install time (dpkg and rpm can handle this, or check differences as part of your auto-setup script).  If the user/admin changes that file, then load their modified file instead of overwriting it.  (If someone has /etc/ mounted read-only, they should be expected to remount rw during install/update of software.)
- instead of 'tmpfs', store the docker-default file somewhere under /var/run/docker/ (and don't delete it while the containers are running).  Then `docker inspect` and `aa-status` can reference the path to the profile, which can easily be seen by those who care.
- I could submit a feature request to the AppArmor folks requesting a config-file directive for their parser, such that it always caches the plaintext versions of all profiles.

**Output of `docker version`:**

```
Client:
 Version:      17.04.0-ce
 API version:  1.28
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.04.0-ce
 API version:  1.28 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   4845c56
 Built:        Mon Apr  3 18:07:42 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

**Output of `docker info`:**

```
Containers: 3
 Running: 3
 Paused: 0
 Stopped: 0
Images: 47
Server Version: 17.04.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 71
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary:
containerd version: 422e31ce907fd9c3833a38d7b8fdd023e5a76e73
runc version: 9c2d8d184e5da67c95d601382adf14862e4f2228
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-75-generic
Operating System: Ubuntu 16.04.2 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 3.859GiB
Name: [redacted]
ID: M2RC:GOLE:6JNO:CRMB:XEO7:FBBN:PNYC:BEUQ:PZUH:L7FN:SZI6:4JPR
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Http Proxy: [redacted]
Registry: https://index.docker.io/v1/
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
VM/Instance on OpenStack (13.1)

**Comment by thaJeztah** [comment]
/cc @n4ss @justincormack wdyt? Perhaps an "export" option to get the profile?

fwiw, I don't think that this is desirable;

>  If the user/admin changes that file, then load their modified file instead of overwriting it. 

As it no longer allows updating the profile when docker is updated, and also could mean that a modified profile is loaded without the user being aware of it

**Comment by tych0** [comment]
Yes, it is quite annoying that the apparmor utilities require a path. FWIW,
```
- instead of 'tmpfs', store the docker-default file somewhere under /var/run/docker/ (and don't delete it while the containers are running). Then docker inspect and aa-status can reference the path to the profile, which can easily be seen by those who care.
```
seems like the best option to me (and then simply always overwrite it before loading).

We could support putting custom profile content in if we generate one profile per container. IIRC, this also has some small security benefit, although I don't remember exactly what.

**Comment by cyphar** [comment]
> After v1.13, Docker now generates docker-default in tmpfs, uses apparmor_parser to load it into kernel, then deletes the file. All of the AppArmor utils (aa-* on Ubuntu) expect a file parameter, and /sys/kernel/security/apparmor/policy/profiles/* only has cached binaries.

For context, the reason we no longer write to `/etc/apparmor.d` is because certain distributions want to make the rootfs read-only. A specific instance of this is SUSE's Kubernetes distribution. The reason I made Docker delete the file after generating it is so that we don't waste space in `/tmp` (and if we made assumptions about the file not changing, that feels quite dodgy).

However, what **would** be good is if we improve the AppArmor profile code in a similar way to the seccomp profile code -- that there is a configuration file in `/etc/docker` which describes the default profile. That way it won't be opaque but also won't require us to keep around temporary copies of the profile. WDYT?

**Comment by spawnflagger** [comment]
Hi guys, thanks for considering this.

>>If the user/admin changes that file, then load their modified file instead of overwriting it.

>As it no longer allows updating the profile when docker is updated, and also could mean that a modified profile is loaded without the user being aware of it

>  is because certain distributions want to make the rootfs read-only

There is a fundamental difference between _install time_ and _run time_.  During install time, it's the root user that is modifying the system, and it's necessary (for most daemon-type software) to make changes in /etc.  Run time (including `/etc/init.d/something start`, reboots, `service start something`, `systemctl up`) - those should never need to _write_ anything to `/etc`, only read from it.  All of those runtime record-keeping changes should be somewhere under `/var` (such is common with PID files).

There are many packages on linux where the manager (dpkg/apt, rpm/yum) will prompt the admin if there is an updated config file and it detects difference on the filesystem- usually something like:  `Keep/Replace/Show Diff? [default Keep]:`  and then it puts the new file in the same folder with an extra extension like .rpmnew or .dpkg-dist.  (the interactive prompts can be overridden via extra parameters)
Here's a Debian article with more details: https://raphaelhertzog.com/2010/09/21/debian-conffile-configuration-file-managed-by-dpkg/

I used the Docker install script via `curl -sSL https://get.docker.com/ | sh`, and that could also be modified to behave like dpkg/rpm, at least for conf-files.


>  that there is a configuration file in `/etc/docker` which describes the default profile

If it's in /etc (either `/etc/docker/` or `/etc/apparmor.d/`) then if the user makes changes to that default, then those changes should be honored by docker, and that modified file should be loaded, rather than generating one from an internal template.  If you want something that is purely informational, then it should be part of the documentation (installation directory varies by distro & package), or a man page (although this is easy to get out-of-sync with reality - I prefer just looking at the file that's actually being used in /etc)

> so that we don't waste space in `/tmp`

These files are only a few kB, so I'm not concerned about the size; however many of them piling up over time can get messy.  `/tmp` also implies "safe to delete, as long as no process has it open (via `lsof`)"

**Comment by cyphar** [comment]
@spawnflagger 

> There is a fundamental difference between install time and run time. During install time, it's the root user that is modifying the system, and it's necessary (for most daemon-type software) to make changes in /etc.

I'm aware of that (and I also am aware of how distributions deal with configuration files). What I'm saying is that Docker used to write to `/etc/apparmor.d` during **run time**, which is why it doesn't do that anymore (and the change was necessary because it used to be the case that Docker would attempt to write to paths such as `/etc/apparmor.d/` that certain setups had set to read-only).

> If it's in /etc (either /etc/docker/ or /etc/apparmor.d/) then if the user makes changes to that default, then those changes should be honored by docker, and that modified file should be loaded, rather than generating one from an internal template.

My suggestion was that we don't have an internal template anymore and that we make it external (and store it in `/etc/docker/apparmor.conf` or something). This is similar to how the seccomp profile used to be internal and is now an external file (`/etc/docker/seccomp.json` IIRC).

> These files are only a few kB

Sure but if you have to regenerate a new profile each time it's not loaded in the system, and then if you combine that with lots of Docker reboots on a dev machine (and certain distributions don't clear `/tmp` on reboot) things start to get a bit hairy. All of that being said, I would prefer if we solve a more general problem here (make the internal profile more visible) rather than trying to retrofit introspection (write the profile to some temporary directory that only people who have read this issue would know about).

**Comment by spawnflagger** [comment]
@cyphar 
> My suggestion was that we don't have an internal template anymore and that we make it external (and store it in /etc/docker/apparmor.conf or something). This is similar to how the seccomp profile used to be internal and is now an external file (/etc/docker/seccomp.json IIRC) -- see issue #26276. Would this work for you?

I think that would be a good step just for the sake of consistency between how Docker handles apparmor and seccomp.

But consider the scenario where there's a bug in the template file or parser- what gets loaded in AppArmor might not be what was intended, and there'd still be no way of knowing what's actually loaded in AppArmor policy (once removed from tmpfs, no way to see plaintext).  Having the file in `/var/run/docker/some_per-container_subdir/aa.profile` then `aa-status` would report the full path, which can be checked.  And it would automatically be deleted when the container is killed.

**Comment by aauren** [comment]
Is there any further progress on this issue? I have to say it's incredibly frustrating when some part of the apparmor profile doesn't work correctly, because as this issue suggests there is absolutely no way to view the profile that Docker is using. Furthermore, all of the apparmor utilities require the original text of the profile in order to modify them in any way.

We recently ran into: https://github.com/moby/moby/issues/36809 and because there is no profile accessible to me, the only way that I have found to kill the containers that are spun up on affected hosts is to redeploy the system.

At the very least an export option would be handy, but if not that, then I would agree with what @spawnflagger suggested, that the file should be kept in tmpfs while the container is running and that the full location of of the applied apparmor profile be logged by docker so that it can easily be tracked down and interrogated.

**Comment by cyphar** [comment]
@aauren I would argue that having access to the *output* profile (which has already been loaded by Docker) will not be as useful as having access to the *input* profile (which will be loaded by Docker in all future invocations) for solving problems like #36809 -- which just requires a one-line fix to the internal AppArmor profile.

I am willing to work on having an `/etc/docker/apparmor.conf` (or whatever the file will be called), as I think it would solve most of the issues. If you *really* would like to know what the output looks like we could make it something like `/etc/docker/apparmor.conf.in` (for the template) and then `/var/run/docker/apparmor.conf` (for the output).

**Comment by aauren** [comment]
@cyphar Thanks for your quick reply!

I think that it would be best to have both. It would be nice to have access to the template so that you could fix problems that you found in a way that docker would actually use them going forward, but it would also be nice to have access to the rendered output of the template (the text that was sent to apparmor) so that you could manipulate apparmor directly.

Almost every interaction that you have with apparmor is contingent on having the profile file that was fed into apparmor. Even tools like `aa-logprof`, as I found out today, won't work unless they have access to the existing profile file that was used to construct the enabled profile.

So for now I would take whatever you guys are willing to give me, but ultimately it would be nice to have access to both the template file and the rendered output that docker sent to apparmor.

**Comment by aauren** [comment]
@cyphar Just following up with this, do you think that this is something that is relatively easy and could be done soon? Or should we pursue a wrapper script like @n4ss suggested if we needed this functionality some time this week?

**Comment by cyphar** [comment]
I would definitely recommend writing the wrapper. I'm working on the patch at the moment and you immediately hit problems because the `--security-opt seccomp` and `--security-opt apparmor` are completely different semantically. In the former, you provide a policy file (so having a `default.json` is useful). In the latter, you provide the name of the **already loaded** AppArmor profile (so having the default profile visible is okay but still a bit ugly -- also generating the default profile at compile-time with `go generate` will be a bit more difficult because it has distro-specific templating code). I'm working on it, there's just quite a few bumps to deal with.

**Comment by aauren** [comment]
Thanks for the information and for your work on this @cyphar !

**Comment by dalguete** [comment]
Following @n4ss suggestions, I created a wrapper around **apparmor_parser** to obtain the profile created by Docker. It's super simple, so don't try it in a stable/prod env. A VM will do the trick. Once done, don't forget to set **apparmor_parser** back to original.

https://gist.github.com/dalguete/19a9d984e8a749eefaab7ec9aea6be01

Steps I followed to have this working:
1. Uninstall Docker.
2. Backup original **apparmor_parser**
   `sudo cp $(which apparmor_parser) $(which apparmor_parser).original`
3. Create the **apparmor_parser** wrapper using code in gist. Don't forget to make that file executable
   `sudo chmod +x <path/to/apparmor_parser>`
4. Reboot (couldn't find a way to reload apparmor profiles without rebooting)
5. Install Docker.
6. See the profile loaded in **/tmp/apparmor-profile-docker-default**.
7. Restore **apparmor_parser**
   ```
   sudo rm $(which apparmor_parser)
   sudo mv <path/to/apparmor_parser>.original <path/to/apparmor_parser>
8. Reboot.

**Comment by vrothberg** [comment]
@jodumont, neither CaaSP nor openSUSE Kubic are abandoned, see https://kubic.opensuse.org/. The download link just doesn't exist anymore as Kubic is rolling now much faster as before and is in line with openSUSE Tumbleweed. The new download links include the snapshot date.

SUSE CaaS Platform is used in production pretty much around the globe now, while openSUSE Kubic is currently under some rather bigger architectural changes as described in this blog post: https://kubic.opensuse.org/blog/2018-08-09-newdirection/

**Comment by jodumont** [comment]
@vrothberg glad to ear, because, at least for me it was a sad impression

**Comment by disconnect3d** [comment]
Bump. It's two years since the issue appeared and it seems we still can't fetch the profile :(.

**Comment by AkihiroSuda** [comment]
PR: https://github.com/moby/moby/pull/39923

**Comment by dannysauer** [comment]
> neither CaaSP nor openSUSE Kubic are abandoned

That was true in 2018, but for People Of The Future: it is no longer true.  CaaSP was basically killed in 2020 (when SUSE quietly fired most of the CaaSP engineers to boost numbers right before formally acquiring Rancher). The last CaaSP release EOL'd in 2021.  Kubic is also now officially wound down as of a couple months ago.

There are still read-only root systems, of course.  But why does read-only root have anything to do with this?  The profile is static for a given release, so it should only need to be written (in other words "installed") at *install* or upgrade time.  I can't imagine a reasonable situation where an admin has sufficient permissions to install binaries **and** to reload apparmor, but I somehow doesn't have permission to install an apparmor config file.  When we made SLES into CaaSP, one of the things that happened was that we ran `transactional-update install docker` to install the docker package.  At install time, the root FS is very much not read-only; changes are written to a snapshot which then applies after reboot.  Redhat's implementation works basically the same way - you install into a readable copy of the FS which applies (or is reverted) later.  There's actually no problem at all writing to /etc/apparmor.d on either system at the time of installation/upgrade.

So it seems like this issue mostly happened to solve a problem which didn't actually exist. :/  The real problem is just packaging - how can the default apparmor profile be packaged with the binary so they get installed.  That already happens with the rest of the config files, right?

**Comment by cyphar** [comment]
@dannysauer 
>  There's actually no problem at all writing to /etc/apparmor.d on either system at the time of installation/upgrade.

This doesn't affect Docker's apparmor profiles. Docker loads them directly from an in-memory copy (which cannot be modified by admins). The issue which this issue is about (and I tried to solve with #41442) is that you cannot provide a hotfix for AppArmor-related issues (we've had both security issues and regressions related to AppArmor where being able to provide a basic fix to the default AppArmor profile would've been nice).

Also Docker doesn't write the profile to `/etc/apparmor.d` anymore specifically because of read-only root filesystems (this was done at runtime as a way for users to be able to see what the profile was -- but this didn't work on read-only root filesystems because the write would fail and Docker was using the file as the source file of the `apparmor_parser` invocation).

#41442 (which needs to be rebased) will fix all of the known outstanding issues with AppArmor profiles under Docker (most notably the issue that Docker does not reload the `docker-default` profile after a restart meaning that even after a security update your containers won't get the new profile without a reboot).

**Comment by dannysauer** [comment]
@cyphar
> Docker doesn't write the profile to /etc/apparmor.d anymore specifically because of read-only root filesystems

That's what I was talking about. So-called "read only root" filesystem *are* writable at install time. The Docker config files in /etc somehow still get created; those weren't moved to some transient ramfs. There's really no reason for the profile to be written at runtime, since it's static for the lifetime of a given docker version.

**Comment by cyphar** [comment]
> There's really no reason for the profile to be written anywhere at runtime, since it's static for the lifetime of a given docker version.

#41442 makes the profile be something you can generate and package (a-la `/etc/docker/seccomp.json`). The issue is that certain rules are applied based on the AppArmor version so there is some runtime logic needed when dealing with the apparmor profile -- so yes, it does need to be generated at runtime.

**Comment by Crypto-Spartan** [comment]
Bump!

It seems it's been a while since anyone has taken a look at this, but I'm hoping there may still be a way to fix this. It would be _very_ useful to be able to export/view the `docker-default` AppArmor profile. 

What brought me here is an issue on OpenMediaVault with the Proxmox kernel where AppArmor prevents docker access to the network socket:
- https://forum.openmediavault.org/index.php?thread%2F47281-docker-not-working-after-change-to-pve-kernel%2F=
- https://forum.openmediavault.org/index.php?thread%2F46112-docker-not-working-since-omv-upgrade%2F=&postID=337053#post337053

Currently, the known fix is to completely disable/uninstall AppArmor. I see this as a less than ideal solution. I'd love to be able to troubleshoot the issue, but I ran into the same thing that brought many others to this thread for 5+ years - I wasn't able to actually find out what the `docker-default` AppArmor profile was.

I did come across these:
- https://test-dockerrr.readthedocs.io/en/latest/security/apparmor/
- https://raw.githubusercontent.com/moby/moby/master/profiles/apparmor/template.go

as well as these, which are linked above:
- https://gist.github.com/dalguete/19a9d984e8a749eefaab7ec9aea6be01
- https://gist.github.com/disconnect3d/d578af68b09ab56db657854ec03879aa

Out of all 4, [this](https://gist.github.com/disconnect3d/d578af68b09ab56db657854ec03879aa) appears to be the most promising, but it's still 4 years old, I have no idea if it's out of date. My next plan is to run @dalguete 's [script](https://gist.github.com/dalguete/19a9d984e8a749eefaab7ec9aea6be01) when I have time. Hopefully we can get eyes on this once again!

**Comment by cyphar** [comment]
@Crypto-Spartan 

The complete apparmor profile is defined in [profiles/apparmor/template.go](https://raw.githubusercontent.com/moby/moby/master/profiles/apparmor/template.go), it's just written using Go's `text/template`. We could make a script to generate it (like we do for seccomp) but the most ideal solution IMHO would be to make it configurable -- which is what #41442 does (I need to go rebase it, though from memory there seemed to be little interest in merging it).

**Comment by Crypto-Spartan** [comment]
On my system, apparmor is actually installed. I was able to add 
```
security_opt:
    - apparmor:unconfined
```
which was able to get the container running.

**Comment by magnus-longva-bouvet** [comment]
Is there any place on the Internet where I can see a plaintext copy of the docker-default AppArmor profile? I would be satisfied by one for containerd, moby or podman as well.

**Comment by xuxucode** [comment]
A js script to get latest `docker-default` AppArmor profile by parsing [template.go](https://raw.githubusercontent.com/moby/moby/v25.0.3/profiles/apparmor/template.go):
https://gist.github.com/xuxucode/e365d85cf2cc4e8a844b25b55e55a89f#file-apparmor-docker-default-js

```sh
# run script
node apparmor-docker-default.js
```
Output on Ubuntu:
```
#include <tunables/global>

profile docker-default flags=(attach_disconnected,mediate_deleted) {
  #include <abstractions/base>

  network,
  capability,
  file,
  umount,
  # Host (privileged) processes may send signals to container processes.
  signal (receive) peer=unconfined,
  # dockerd may send signals to container processes (for "docker kill").
  signal (receive) peer=unconfined,
  # Container processes may send signals amongst themselves.
  signal (send,receive) peer=docker-default,

  deny @{PROC}/* w,   # deny write for all files directly in /proc (not in a subdir)
  # deny write to files not in /proc/<number>/** or /proc/sys/**
  deny @{PROC}/{[^1-9],[^1-9][^0-9],[^1-9s][^0-9y][^0-9s],[^1-9][^0-9][^0-9][^0-9/]*}/** w,
  deny @{PROC}/sys/[^k]** w,  # deny /proc/sys except /proc/sys/k* (effectively /proc/sys/kernel)
  deny @{PROC}/sys/kernel/{?,??,[^s][^h][^m]**} w,  # deny everything except shm* in /proc/sys/kernel/
  deny @{PROC}/sysrq-trigger rwklx,
  deny @{PROC}/kcore rwklx,

  deny mount,

  deny /sys/[^f]*/** wklx,
  deny /sys/f[^s]*/** wklx,
  deny /sys/fs/[^c]*/** wklx,
  deny /sys/fs/c[^g]*/** wklx,
  deny /sys/fs/cg[^r]*/** wklx,
  deny /sys/firmware/** rwklx,
  deny /sys/devices/virtual/powercap/** rwklx,
  deny /sys/kernel/security/** rwklx,

  # suppress ptrace denials when using 'docker ps' or using 'ps' inside a container
  ptrace (trace,read,tracedby,readby) peer=docker-default,
}
```

**Comment by Gunni** [comment]
I found a way to disable apparmor on docker.

Just set the environment variable `container` to any value.

`$ systemctl edit docker`

```systemd
[Service]
Environment=container="disable apparmor"
```

https://github.com/containerd/containerd/blob/8317959018015f6a1756ec8cd08be1093fd630a2/pkg/apparmor/apparmor_linux.go#L37

It entirely disabled apparmor for docker and does not even try to load the docker-default apparmor profile.

Good thing I found that, i was going to have to diable apparmor on the host if this wasn't a thing.

Because when I try to stop docker containers, i get this.
```audit
AVC apparmor="DENIED" operation="signal" class="signal" profile="docker-default" pid=2965 comm="runc" requested_mask="receive" denied_mask="receive" signal=kill peer="/usr/sbin/runc"
```

**Comment by rgl** [comment]
FWIW, I've created yet another less-than-half-baked hack at https://github.com/rgl/moby-dockerd-docker-default-apparmor-profile/tree/docker-default-apparmor-profile-cmd. I've modified `dockerd` to have a `docker-default-apparmor-profile` command to dump the `docker-default` profile to `stdout`. Its used as:

```bash
wget https://github.com/rgl/moby-dockerd-docker-default-apparmor-profile/releases/download/v28.0.4-docker-default-apparmor-profile/dockerd-docker-default-apparmor-profile.tgz
tar xf dockerd-docker-default-apparmor-profile.tgz
sudo ./dockerd docker-default-apparmor-profile >docker-default.apparmor
cat docker-default.apparmor
```

**Comment by bduffany** [comment]
Would the docker team be open to a PR that adds an option like `docker info --apparmor` or maybe `docker info --apparmor-profile`, which would display the apparmor profile on stdout? I To me, this seems to be a fairly natural/intuitive place for it, since `docker info` already displays information about the apparmor profile, but it just shows the string "apparmor" under "Security Options" section, and nothing else.

My use case is generating a custom profile that _overrides_ docker's default profile: I want to delete the line `deny mount` but keep all the other lines, since I need `mount` perms for a nested-container use case. Today, my options for doing an override in this way are basically
- (A) Use this [workaround](https://github.com/moby/moby/issues/33060#issuecomment-1948635934) from earlier in the thread that attempts to reconstruct the profile based on the Go template - this script works, but it could break at any time if Docker changes their template.go file (in fact - it's already broken because the profile was moved to the moby/profiles repo). I could instead do a one-time run of this script, but that means I'll never get security updates.
- (B) Disable AppArmor, which is not ideal

(My needs would also be addressed if there were a way to load multiple profiles, where one can override another, but I think that'd require kernel changes, which seems much more difficult)

**Comment by AkihiroSuda** [comment]
> `docker info --apparmor-profile`

Maybe `docker apparmor inspect`?

The convention is already implemented in nerdctl: https://github.com/containerd/nerdctl/blob/main/docs/command-reference.md#apparmor-profile-management

```console
$ nerdctl apparmor
Manage AppArmor profiles

Usage: nerdctl apparmor [flags]

Commands:
  inspect  Display the default AppArmor profile "nerdctl-default". Other profiles cannot be displayed with this command.
  load     Load the default AppArmor profile "nerdctl-default". Requires root.
  ls       List the loaded AppArmor profiles
  unload   Unload an AppArmor profile. The target profile name defaults to "nerdctl-default". Requires root.

Flags:
  -h, --help   help for apparmor

See also 'nerdctl --help' for the global flags such as '--namespace', '--snapshotter', and '--cgroup-manager'.
```

<details><summary><code>nerdctl apparmor inspect</code></summary>

```console
$ nerdctl apparmor inspect


#include <tunables/global>


profile nerdctl-default flags=(attach_disconnected,mediate_deleted) {

  #include <abstractions/base>


  network,
  capability,
  file,
  umount,
  # Host (privileged) processes may send signals to container processes.
  signal (receive) peer=unconfined,
  # runc may send signals to container processes.
  signal (receive) peer=runc,
  # crun may send signals to container processes.
  signal (receive) peer=crun,
  # Manager may send signals to container processes.
  signal (receive) peer=unconfined
,
  # Container processes may send signals amongst themselves.
  signal (send,receive) peer=nerdctl-default,

  # https://github.com/containerd/nerdctl/issues/2730
  signal (receive) peer=/usr/local/bin/rootlesskit,


  deny @{PROC}/* w,   # deny write for all files directly in /proc (not in a subdir)
  # deny write to files not in /proc/<number>/** or /proc/sys/**
  deny @{PROC}/{[^1-9],[^1-9][^0-9],[^1-9s][^0-9y][^0-9s],[^1-9][^0-9][^0-9][^0-9]*}/** w,
  deny @{PROC}/sys/[^k]** w,  # deny /proc/sys except /proc/sys/k* (effectively /proc/sys/kernel)
  deny @{PROC}/sys/kernel/{?,??,[^s][^h][^m]**} w,  # deny everything except shm* in /proc/sys/kernel/
  deny @{PROC}/sysrq-trigger rwklx,
  deny @{PROC}/mem rwklx,
  deny @{PROC}/kmem rwklx,
  deny @{PROC}/kcore rwklx,

  deny mount,

  deny /sys/[^f]*/** wklx,
  deny /sys/f[^s]*/** wklx,
  deny /sys/fs/[^c]*/** wklx,
  deny /sys/fs/c[^g]*/** wklx,
  deny /sys/fs/cg[^r]*/** wklx,
  deny /sys/firmware/** rwklx,
  deny /sys/devices/virtual/powercap/** rwklx,
  deny /sys/kernel/security/** rwklx,

  # allow processes within the container to trace each other,
  # provided all other LSM and yama setting allow it.
  ptrace (trace,tracedby,read,readby) peer=nerdctl-default,
}
```

</details>

**Comment by carsonl** [comment]
> FWIW, I've created yet another less-than-half-baked hack at https://github.com/rgl/moby-dockerd-docker-default-apparmor-profile/tree/docker-default-apparmor-profile-cmd. I've modified `dockerd` to have a `docker-default-apparmor-profile` command to dump the `docker-default` profile to `stdout`. Its used as:
> <--snip-->

I've updated this hack from @rgl to align with the latest upstream version: https://github.com/moby/moby/compare/master...carsonl:moby-dockerd-docker-default-apparmor-profile:moby-dockerd-docker-default-apparmor-profile

Same commands as above, except for the wget, which would be:
```
wget https://github.com/carsonl/moby-dockerd-docker-default-apparmor-profile/releases/download/v29.1.3/dockerd-docker-default-apparmor-profile.tgz
tar xf dockerd-docker-default-apparmor-profile.tgz
sudo ./dockerd docker-default-apparmor-profile >docker-default.apparmor
cat docker-default.apparmor
```

**Comment by carsonl** [comment]
Oh good, what a waste of time, the profile it outputs is effectively identical.
